Totally change how "dp deploy" works...

ian-ross opened the pull request:
Before this commit, `dp` created a deployment tarball from the **current working directory**, irrespective of what branch you told it to deploy, and irrespective of any uncommitted or even unstaged changes in the working directory.  This commit attempts to make the behaviour of `dp deploy` a little more sane.

The new `dp deploy` does the following:

- Clones the Git repository where `dp deploy` is run on the remote machine.  The clone appears in the `versions/<date>` directory for the application being deployed.  (The remote machine thus needs to have the appropriate SSH keys in place to allow this to work.)

- Sets up Git submodules and checks out the requested branch.

- Uses the `.dp/compile` script to build the application.

- Creates symbolic links as required.

This procedure means that the deployments ALWAYS happen from a well-defined branch in the Git repository.

saschat commented:
I did not look at the code yet, but re the description: I don't think we want this SSH key requirement for the target machine. We did not want it for the memcachier machines and the same goes for monitor machines. We should minimize the requirement to spread private keys around.

alevy commented:
I agree with @saschat. Is this all stemming basically from the fact that `git archive` doesn't deal with submodules properly?

alevy commented:
If so, what about something like this instead? <https://github.com/meitar/git-archive-all.sh>

dterei commented:
So the current `dp deploy` has two modes and supports choosing which one to use:

* `DIR` - deploy the current directory and all it's sub-directories.
* `GIT` - deploy the entire git repository.

Can you be clear which or if both deployments modes has the bug?

I wrote the support for both as when using a monorepo, it's pretty unsatisfying the `dp` when using `GIT` mode copies the entire repository to the machine when you may need to only copy a fraction of it.

I haven't looked at the code but it sounds like you've dropped `DIR` mode and improved `GIT` mode?

ian-ross commented:
@saschat I've now changed the way this works so that it clones the repository locally to do all the necessary setup.  That means there are no SSH keys needed on the remote machine.  The messaging is also a little better now, I think.

@alevy I don't know if it's to do with submodules per se, but the main problem we found is that `dp` just doesn't deploy the branch you tell it to -- it deploys whatever is currently checked out in your working directory plus any uncommitted and/or unstaged changes.  The right way to deal with this would seem to be to use Git to set up a clone of the repository with exactly the versions of everything you want and all submodules set up, then to ship that to the remote machine for building.  This also fixes the problem we had with building the monitor where there's a makefile goal to generate a `version.go` file that depends on being run inside a valid Git repo.  Previously, `dp` didn't include the `.git` directory in the tarball it created, so that didn't work.  It now works correctly.

@dterei Basically yes, I dropped the `DIR` mode and changed the way that the `GIT` mode works so that it really does deploy the branch/tag you tell it to.  The only places we were using the `DIR` mode were for the New Relic plugin, and for the monitor, and in the case of the monitor, that's now the wrong thing to use, because the monitor build process now relies on being run in a real Git repository so that it can extract version information about what it's building.  As for which deployment mode had the bug, it was both of them: they both just did a `git ls-files` and pull files from the working directory at whatever current state it's in, instead of extracting things from the deployment branch.

My personal feeling about this is that I just want `dp` to be reliable.  We haven't had any screw-ups from this in production deployments yet because Sascha and I were *excessively* careful in the first one we did the other day.  (I think we were both slightly terrified of messing things up!)  I do know that Sascha has accidentally deployed uncommitted changes to staging before now though, and when we discovered this behaviour doing deployments to production the other day, we were a bit shocked...  I don't mind if it takes a while to run a deployment, I just want it to deploy the thing I believe that it's deploying.  The main time sink is uploading the tarball to the remote machine, and I've changed the way that happens so that there's a progress indicator, which removes most of the frustration -- for me, at least!

ian-ross commented:
I'm going to unilaterally merge this, since it seems a good idea to have a working `dp` that does what we think it should do...  We can talk about optimisations and other changes later.

saschat commented:
Sorry, forgot about this. Looks good to me!

saschat commented:
Btw. this fixes #18
